### PR TITLE
chore: Bump curve25519-dalek to v4, sha3 to v0.10

### LINKED ISSUE #647: chore: bump sha3 version
#### Problem

discussed with Sam earlier and we all agree that we should bump sha3 to the latest but we will need to update code base for some API breaking changes 😢

### PR DESCRIPTION
**Pull request by Xuanwo**
#### Problem

> discussed with Sam earlier and we all agree that we should bump sha3 to the latest but we will need to update code base for some API breaking changes

#### Summary of Changes

Fixes https://github.com/anza-xyz/agave/issues/647

This PR fix #647 by:

- Bumping curve25519-dalek to v4
- Bumping sha3 to v0.10

Follow-up tasks:

- Bumping ed25519-dalek to v2 so we can remove all depends of curve25519-dalek v3

### CODE REVIEW

File: zk-token-sdk/src/curve25519/scalar.rs [hunk 1]
```diff
@@ -18,7 +18,7 @@ mod target_arch {
         type Error = Curve25519Error;
 
         fn try_from(pod: &PodScalar) -> Result<Self, Self::Error> {
-            Scalar::from_canonical_bytes(pod.0).ok_or(Curve25519Error::PodConversion)
+            Option::from(Scalar::from_canonical_bytes(pod.0)).ok_or(Curve25519Error::PodConversion)
```
Comment (Xuanwo) [inline zk-token-sdk/src/curve25519/scalar.rs:21]: API like `Scalar::from_canonical_bytes` returns `CtOption` which doesn't have `ok_or` API. Do you have better ideas?

File: zk-token-sdk/src/curve25519/ristretto.rs [hunk 2]
```diff
@@ -64,6 +64,7 @@ mod target_arch {
 
         fn try_from(pod: &PodRistrettoPoint) -> Result<Self, Self::Error> {
             CompressedRistretto::from_slice(&pod.0)
+                .map_err(|_| Curve25519Error::PodConversion)?
```
Comment (Xuanwo) [inline zk-token-sdk/src/curve25519/ristretto.rs:67]: It is good to omit the error and returns `Curve25519Error::PodConversion` directly?

**Comment by samkim-crypto** [comment]
Thanks so much for the contribution! I am starting to look at this (sorry for the delay 🙏 ), but it seems like there is a duplicate PR https://github.com/anza-xyz/agave/pull/513 that also upgrades the curve25519-dalek version. Since the other one is an ealier PR, let's wait until how it turns out first.

**Comment by joncinque** [comment]
Closing since this was done with #1693. Thanks for your contribution anyway!

**Pull request closed without merging**